list.man
.TH LIST 3  "September 22, 1991"
.UC 6
.SH NAME
list_init, list_mvprev, list_mvnext, list_mvfront, list_mvrear, list_front, list_curr, list_rear, list_insert_before, list_insert_after, list_remove_front, list_remove_curr, list_remove_rear, list_size, list_empty, list_traverse, list_free \-- generic linked-list routines
.SH SYNOPSIS
.PP
cc [
.I flags 
]
.IR files
.B -llist 
[ 
.IR libraries
]
.nf
.PP
.ft B
\#include <stdint.h>
\#include <stddef.h>
\#include <list.h>
.PP
.ft B
typedef void (*list_deallocator_t)(void *data);
typedef list_boolean_t (*list_comparator_t)(void *userdata, void *listdata); 
.PP
.ft B
LIST *list_create()
void list_destroy(LIST *list, list_deallocator_t func);
.PP
.ft B
uint32_t list_size(LIST *list);
list_boolean_t int list_empty(LIST *list);
.PP
.ft B
void *list_front(LIST *list);
void *list_curr(LIST *list);
void *list_rear(LIST *list);
.PP
.ft B
LIST_ELEMENT *list_element_front(LIST *list);
LIST_ELEMENT *list_element_curr(LIST *list);
LIST_ELEMENT *list_element_rear(LIST *list);
.PP
.ft B
LIST *list_mvprev(LIST *list);
LIST *list_mvnext(LIST *list);
LIST *list_mvfront(LIST *list);
LIST *list_mvrear(LIST *list);
LIST *list_setcurr(LIST *list, LIST_ELEMENT *element);
.PP
.ft B
void *list_insert_before(LIST *list, void *data, size_t bytes);
void *list_insert_after(LIST *list, void *data, size_t bytes);
.PP
.ft B
void *list_remove_front(LIST *list);
void *list_remove_curr(LIST *list);
void *list_remove_rear(LIST *list);
.PP
.ft B
int list_traverse(LIST *list, void *userdata, list_comparator_t func, int opts);
.PP
.SH DESCRIPTION
These routines provide generic manipulation of (potentially)
multiple linked lists.  Each list can hold arbitrarily sized elements, with
individual elements within a list varying in size.  It is the
programmer's responsibility to account for such differences.

Lists are referred to by variables declared as LIST *; the type LIST
is defined in <list.h> to be an opaque structure.  Internally each descriptor
maintains references to the front, rear, and current elements, as well as the
size (or length) of the list.  Various routines operate relative to the
current element.
.PP
.IR list_create
creates and returns an opaque pointer to the list. This pointer is used for
all subsequent list operations. 
.PP
.IR list_destroy
destroys the 
.IR list ,
by deallocating the list applying the user-supplied function
.IR list_deallocator_t
.IR func
to the data portion of each element remaining in the list.  If
.IR func 
is (void *)LIST_DEALLOC, then the package will apply its own deallocation
routine.  This, however, should only be done if the package has been
responsible for data element allocation, i.e., the insert routines
have been invoked with
.IR bytes
greater than 0.  If
.IR func
is (void *)LIST_NODEALLOC, no per-element deallocation will be performed.
.PP
.IR list_size
returns the size (or number of elements) of
.IR list
as an unsigned integer.
.PP
.IR list_empty
returns 1 (LIST_TRUE) if
.IR list
is empty, 0 (LIST_FALSE) otherwise.
.PP
.IR list_front ,
.IR list_curr ,
and
.IR list_rear 
return pointers to the front, current, or rear element in
.IR list .
.PP
.IR list_element_front ,
.IR list_element_curr ,
.IR list_element_rear ,
return an opaque LIST_ELEMENT pointer.  This pointer can only be used to
directly set the list curr pointer with
.IR list_setcurr .
Useful for keeping a back pointer to an array element to avoid list traversal. 
.PP
.IR list_mvprev ,
.IR list_mvnext ,
.IR list_mvfront ,
.IR list_mvrear ,
move to the previous, next, front, or rear element in
.IR list ,
and return the modified list descriptor.  Movement previous or next is relative
to the current element.
.PP
.IR list_insert_before
and
.IR list_insert_after
insert an element, pointed to by
.IR data
and of size
.IR bytes ,
into
.IR list ,
either before or after the current element.  The newly inserted element is
then considered the current element.  Both routines return a pointer to the
newly inserted data.  If
.IR bytes
is greater than 0, then data is copied into the list, otherwise
only the reference
.IR data
is copied into the list.  This allows the user to determine the
memory allocation policy.
.PP
.IR list_remove_front ,
.IR list_remove_curr ,
and
.IR list_remove_rear 
remove the front, current, or rear element from
.IR list 
and return a pointer to the removed data.
The current element is then set to the next element (prior to the remove) in
the list, unless the element at the rear of the list was removed.  In this
case, the current element is set to the previous element (prior to the
remove).
.PP
.IR list_traverse
traverses
.IR list
according to options 
.IR opts ,
calling
.IR list_comparator_t
.IR func
at each element encountered in the traversal, until
.IR func
returns 0 (FALSE) or the extent of the list is reached.  This routine
can be used to search for an item or print the contents of the list,
for example.  See the section LIST TRAVERSAL for more information.
.SH LIST TRAVERSAL
The behavior of the routine
.IR list_traversal
is controlled by the user-supplied function
.IR func
which is responsible for the actions performed at each element in the
traversal.  In addition, the scope and direction of the
traversal can be specified with the
.IR opts
parameter.
.IR Func
should be declared as follows:
.PP
.RS
list_boolean_t func(void *userdata, void *listdata); 
.RE
.PP
and should return 1 (LIST_TRUE) if the traversal should continue, or 0
(LIST_FALSE) if the traversal should terminate.  The function does not
need to check if the extent of the list has been reached; the return code
from
.IR list_traverse
will indicate the status of the traversal.  At each element encountered
in the traversal,
.IR list_traverse
will invoke
.IR func ,
passing it the two parameters
.IR userdata
and
.IR listdata ;
.IR userdata
being the data pointer provided by the caller in the call to
.IR list_traverse ,
and
.IR listdata
being the data pointer of the current list element in the traversal.  \
For example, by passing the following function to
.IR list_traverse ,
the contents of a list of names could be printed.
.PP
.RS
.nf
list_boolean_t func(void *userdata, void *listdata)
{
    printf("Name=%30s.\\n", (char *)listdata->name);
    return(LIST_TRUE);
}
.fi
.RE
.PP
In this example, the parameter
.IR userdata
is ignored and the function unconditionally returns 1 (LIST_TRUE), but functions
like
.PP
.RS
.nf
list_boolean_t func(void *userdata, void *listdata)
{
    if (strcmp((char *)userdata, (char *)listdata))
       return(LIST_TRUE);
    else
       return(LIST_FALSE);
}
.fi
.RE
.PP
can be used to position the current element pointer prior to insertion, so
as to keep the list ordered.  
.PP
The       direction and scope of the traversal can be controlled by specifying
one or more of the following options:
.PP
.RS
.nf
.ta 1.0i +1.0i +.5i
LIST_FORW	*	traverse forward (next)
LIST_BACK		traverse backwards (prev)
LIST_FRNT	*	start from the front of the
		list (implies LIST_FORW)
LIST_CURR		start from the current element
LIST_REAR		start from the rear element
		(implies LIST_BACK)
LIST_SAVE	*	do not alter the current element
		pointer during the traversal
LIST_ALTR		alter the current element pointer
		during the traversal
.fi
.RE
.PP
The asterisks (*) denote the default values.  These options can be combined
with the logical OR operator, but at least one value must be specified.
For example, specifying
.PP
.RS
LIST_FORW
.RE
.PP
for
.IR opts
would request a traversal forwards from the current position,
restoring the current element pointer after the traversal, whereas
.PP
.RS
(LIST_BACK | LIST_CURR | LIST_ALTR)
.RE
.PP
would request a traversal backwards from the current position, and would
set the current element pointer to the last element encountered in the
traversal.

It should be noted that
.IR func
should not invoke any of the list routines unless LIST_ALTR has been
specified, since many of the routines act relative to the current
element pointer, which is not modified during a traversal with
LIST_SAVE specified.
.SH MEMORY ALLOCATION
The routines
.IR list_init ,
.IR list_insert_before ,
and
.IR list_insert_after
allocate memory during their execution.  As such, 
.IR list_insert_before
and
.IR list_insert_after
insert a copy of the data into the list when they are invoked with
.IR bytes
greater than 0.  If
.IR bytes
is 0, then only the reference is copied into the list.  This allows
the user to control the memory allocation policy.  Both functions may
fail during memory allocation; see DIAGNOSTICS below for more information.

Note that
.IR list_remove_front ,
.IR list_remove_curr ,
and
.IR list_remove_rear
do not allocate memory for the removing data.  They simply disassociate the
data from the list, and thus return a pointer to data that was previously
allocated by the package.  It is the programmer's responsibility to deallocate
such removed data.

If the user has been responsile for data element storage allocation, i.e. the
insert routines have been called with
.IR bytes
equal to 0, then the user must be responsible for storage deallocation
as well.  A user-supplied deallocation function should be passed to
.IR list_destroy
for this purpose.  The deallocation function should be declared as:
.PP
.RS
void *func(void *data);
.RE
.PP
and will be sent the data element reference of each element in the list
when
.IR list_free
is invoked.  If the package has been responsible for data element allocation,
.IR list_free
can be invoked with (VOID *)LIST_DEALLOC for
.IR func,;
the list package will apply its own deallocation routine, or
(void *)LIST_NODEALLOC if no per-element deallocation is required.  It is the
programmer's responsibility to insure that the memory allocation
policy is applied properly.
.SH DIAGNOSTICS
A NULL returned by
.IR list_init ,
.IR list_insert_before ,
or
.IR list_insert_after
indicates a failure in allocating memory for the new list or element.  See
.IR malloc (3)
for more information.

.IR list_mvprev ,
.IR list_mvnext ,
.IR list_mvfront ,
.IR list_mvrear ,
.IR list_front ,
.IR list_curr ,
.IR list_rear ,
.IR list_remove_front,
.IR list_remove_curr,
and
.IR list_remove_rear
all return NULL if
.IR list
is empty.

.IR list_traverse
returns LIST_EMPTY for an empty list, LIST_EXTENT if an attempt was
made to move beyond the extent of the list, or LIST_OK otherwise.

A core dump indicates a bug ;-)
.SH BUGS
The routines
.IR list_remove_front ,
.IR list_remove_curr ,
.IR list_remove_rear ,
and
.IR list_free
do not physically reclaim storage space, although they do make it
available for reuse.  While this is a function of
.IR free (3),
its application here could be considered a bug.
.SH SEE ALSO
queue(3), stack(3), cache(3)
.SH AUTHOR
Bradley C. Spatz (bcs@ufl.edu), University of Florida.
.PP
Philip Kufeldt


list.c
/* list.c -- a generic list package
 * 
 * Last edited: Tue Jul 28 15:37:24 1992 by bcs (Bradley C. Spatz) on wasp
 *
 * Copyright (C) 1992, Bradley C. Spatz, bcs@ufl.edu
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA  02111-1307, USA.
 *
 * We define the following routines here:
 *
 * LIST *list_create()
 * void  list_destroy(LIST *list, list_deallocator_t func)
 *
 * Simple list size utility routines
 * int   list_empty(LIST *list)
 * int   list_size(LIST *list)
 *
 * Basic list data access routines
 * void *list_front(LIST *list)
 * void *list_curr(LIST *list)
 * void *list_rear(LIST *list)
 *
 * List element access
 * LIST_ELEMENT *list_element_front(LIST *list)
 * LIST_ELEMENT *list_element_curr(LIST *list)
 * LIST_ELEMENT *list_element_rear(LIST *list)
 *
 * Basic list movement routines
 * LIST *list_mvprev(LIST *list)
 * LIST *list_mvnext(LIST *list)
 * LIST *list_mvfront(LIST *list)
 * LIST *list_mvrear(LIST *list)
 * LIST *list_mvcurr(LIST *list, LIST_ELEMENT *element)
 *
 * List manipulation routines
 * void *list_insert_before(LIST *list, void *data, size_t bytes)
 * void *list_insert_after(LIST *list, void *data, size_t bytes)
 * void *list_remove_front(LIST *list)
 * void *list_remove_rear(LIST *list)
 * void *list_remove_curr(LIST *list)
 *
 * int   list_traverse(LIST *list, void *data, list_comparator_t func, int opts)
 */

static char brag[] = "$$Version: list-2.1 Copyright (C) 1992 Bradley C. Spatz\n\
$$Version: list-2.2 Copyright (C) 2021 Philip Kufeldt";

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <stddef.h>
#include "list.h"

static uint32_t list_id=1000; /* list ids are allocated from here */

/* Define an internal structure to describe each element in the list. */
typedef struct list_element {
	uint32_t id; 
	struct list_element *prev;
	struct list_element *next;
	void *data;
} list_element_t;

/* Define an internal structure to describe the list */
typedef struct list {
	uint32_t id;			/* List ID, matches with Element IDs */
	uint32_t size;			/* Number of elements in the list */
	list_element_t *front;	/* Front element of the list */
	list_element_t *rear;	/* Tail element of the list */
	list_element_t *curr;	/* Current element */
} list_t;

/* List creation and destruction routines */
#define SBCAS __sync_bool_compare_and_swap

LIST *
list_create()
{
	list_t *list;

	/* Allocate, initialize, and return a new list. */
	list = (LIST *) malloc(sizeof(list_t));
	list->size = 0;
	list->front = NULL;
	list->rear = NULL;
	list->curr = NULL;

	/* Grab the next id, use compare and swap since list_id is a global */
	list->id = list_id;
	while (!SBCAS(&list_id, list->id, list->id + 1)) {
		list->id = list_id;
	}
	
	return((LIST *)list);
}


void
list_destroy(LIST *ulist, list_deallocator_t func)
{
	void *data;
	list_t *list = (list_t *)ulist;
	
	/* 
	 * Move to the front of the list.  Start removing elements from the
	 * front.  Free up the data element by either applying the user-supplied
	 * deallocation routine or free(3).  When we've gone through all the
	 * elements, free the list descriptor.
	 */
	list_mvfront(list);
	while (! list_empty(list)) {
		data = list_remove_front(list);
		/* Apply either no deallocation function to each node, 
		 * our own, or a user-supplied version.
		 */
		if ((enum list_constants)func != LIST_NODEALLOC) {
			if ((enum list_constants)func == LIST_DEALLOC) {
				free(data);
			} else {
				(*func)(data);
			}
		}
	}

	free(list);
}


/* Simple list size utility routines */
list_boolean_t
list_empty(LIST *ulist)
{
	list_t *list = (list_t *)ulist;

	/* Return 1 if the list is empty.  0 otherwise. */
	return((list->front == NULL) ? LIST_TRUE : LIST_FALSE);
}


uint32_t
list_size(LIST *ulist)
{
	list_t *list = (list_t *)ulist;
	return(list->size);
}


/* Basic list data access routines */
void *
list_front(LIST *ulist)
{
	list_t *list = (list_t *)ulist;
	return((list->front == NULL) ? NULL : (list->front->data));
}


void *
list_curr(LIST *ulist)
{
	list_t *list = (list_t *)ulist;
	return((list->curr == NULL) ? NULL : (list->curr->data));
}


void *
list_rear(LIST *ulist)
{
	list_t *list = (list_t *)ulist;
	return((list->rear == NULL) ? NULL : (list->rear->data));
}


/* List element access */
LIST_ELEMENT *
list_element_front(LIST *ulist)
{
	list_t *list = (list_t *)ulist;
	return((list->front == NULL) ? NULL : ((LIST_ELEMENT *)list->front));
}

 
LIST_ELEMENT *
list_element_curr(LIST *ulist)
{
	list_t *list = (list_t *)ulist;
	return((list->front == NULL) ? NULL : ((LIST_ELEMENT *)list->curr));
}

 
LIST_ELEMENT *
list_element_rear(LIST *ulist)
{
	list_t *list = (list_t *)ulist;
	return((list->front == NULL) ? NULL : ((LIST_ELEMENT *)list->rear));
}


/* Basic list movement routines */
LIST *
list_mvprev(LIST *ulist)
{
	list_t *list = (list_t *)ulist;
	
	/* Move to the previous link, if possible.  Note that the following
	 * compound conditional expression *requires* a short-circuit evaluation.
	 */
	
	if ((list->curr != NULL) && (list->curr->prev != NULL)) {
		list->curr = list->curr->prev;
		return((LIST *)list);
	} else {
		return(NULL);
	}
}


LIST *
list_mvnext(LIST *ulist)
{
	list_t *list = (list_t *)ulist;

	/* Move to the next link, if possible.  Note that the following
	 * compound conditional expression *requires* a short-circuit evaluation.
	 */
	if ((list->curr != NULL) && (list->curr->next != NULL)) {
		list->curr = list->curr->next;
		return((LIST *)list);
	} else {
		return(NULL);
	}
}


LIST *
list_mvfront(LIST *ulist)
{
	list_t *list = (list_t *)ulist;

	/* Move to the front of the list.*/
	list->curr = list->front;
	return((LIST *)list);
}


LIST *
list_mvrear(LIST *ulist)
{
	list_t *list = (list_t *)ulist;

	/* Move to the front of the list.*/
	list->curr = list->rear;
	return((LIST *)list);
}


LIST *
list_setcurr(LIST *ulist, LIST_ELEMENT *uelement)
{
	list_t *list = (list_t *)ulist;
	list_element_t *element = (list_element_t *)uelement;

	/* Sanity check */ 
	if (list->id != element->id)
		return(NULL);
	
	/* Move to the front of the list.*/
	list->curr = element;
	return((LIST *)list);

}


/* Internal List element creation and single element remove routines */
static list_element_t *
list_create_element(void *data, size_t bytes)
{
   list_element_t *new;

   /* Allocate storage for the new node and its data.  Return NULL if
    * unable to allocate.
    */
   new = (list_element_t *) malloc(sizeof(list_element_t));
   if (new == NULL) {
	   return(NULL);
   }

   /* Allocate storage for the data only if requested; i.e. if bytes > 0.
    * Then either copy the data or just the reference into the node.
    */
   if (bytes > 0) {
	   new->data = (char *) malloc(bytes);
	   if (new->data == NULL) {
		   return(NULL);
	   }
	   (void) memcpy(new->data, data, bytes);
   } else {
	   new->data = (char *) data;
   }

   return(new);
}


static void *
list_remove_single(list_t *list)
{
	char *data;

	/* The list has one element.  Easy. */
	data = list->curr->data;
	free(list->curr);
	list->front = list->rear = list->curr = NULL;
	list->size--;
	return (data);
}


/* List manipulation routines */
void *
list_insert_before(LIST *ulist, void *data, size_t bytes)
{
	list_element_t *new;
	list_t *list = (list_t *)ulist;

	/* Allocate storage for the new element and its data.*/
	new = list_create_element(data, bytes);
	if (new == NULL)
		return(NULL);
	new->id = list->id;

	/* Now insert the element before the current, considering the cases:
	 *    1) List is empty
	 *    2) Inserting at front
	 *    3) Otherwise
	 * We handle them directly, in order.
	 */
	if (list->front == NULL) {
		/* The list is empty.  Easy. */
		new->prev = new->next = NULL;
		list->front = list->rear = list->curr = new;
	} else if (list->curr->prev == NULL) {
		/* Inserting at the front. */
		new->prev = NULL;
		new->next = list->curr;
		list->curr->prev = new;
		list->front = new;
	} else {
		/* Otherwise. */
		new->prev = list->curr->prev;
		list->curr->prev->next = new;
		new->next = list->curr;
		list->curr->prev = new;
	}

	list->curr = new;
	list->size++;
	return(new->data);
}


void *
list_insert_after(LIST *ulist, void *data, size_t bytes)
{
	list_element_t *new;
	list_t *list = (list_t *)ulist;

	/* Allocate storage for the new element and its data.*/
	new = list_create_element(data, bytes);
	if (new == NULL)
		return(NULL);
	new->id = list->id;

	/* Now insert the element after the current, considering the cases:
	 *    1) List is empty
	 *    2) Inserting at rear
	 *    3) Otherwise
	 * We handle them directly, in order.
	 */
	if (list->front == NULL) {
		/* The list is empty.  Easy. */
		new->prev = new->next = NULL;
		list->front = list->rear = list->curr = new;
	} else if (list->curr->next == NULL) {
		/* Inserting at the rear. */
		new->next = NULL;
		new->prev = list->curr;
		list->curr->next = new;
		list->rear = new;
	} else {
		/* Otherwise. */
		new->next = list->curr->next;
		new->next->prev = new;
		new->prev = list->curr;
		list->curr->next = new;
	}

	list->curr = new;
	list->size++;
	return(new->data);
}


void *
list_remove_front(LIST *ulist)
{
	list_element_t *temp;
	list_t *list = (list_t *)ulist;
	void *data;

	/* Removing and return front element, or NULL if empty.  If curr
	 * is the front, then curr becomes the next element.
	 */
	if (list->front == NULL) {
		/* List is empty.  Easy. */
		return(NULL);
	} else if (list->front == list->rear) {
		/* List has only one element.  Easy. */
		data = list_remove_single(list);
	} else {
		/* List has more than one element.  Make sure to check if curr
		 * points to the front.
		 */
		data = list->front->data;
		list->front->next->prev = NULL;
		temp = list->front;
		list->front = temp->next;
		if (list->curr == temp)
			list->curr = temp->next;
		free(temp);
		list->size--;
	}

	return(data);
}


void *
list_remove_rear(LIST *ulist)
{
	list_element_t *temp;
	list_t *list = (list_t *)ulist;
	void *data;

	/* Removing and return rear element, or NULL if empty.  If curr
	 * is the rear, then curr becomes the previous element.
	 */
	if (list->front == NULL) {
		/* List is empty.  Easy. */
		return(NULL);
	} else if (list->front == list->rear) {
		/* List has only one element.  Easy. */
		data = list_remove_single(list);
	} else {
		/* List has more than one element.  Make sure to check if curr
		 * points to the rear.
		 */
		data = list->rear->data;
		list->rear->prev->next = NULL;
		temp = list->rear;
		list->rear = temp->prev;
		if (list->curr == temp)
			list->curr = temp->prev;
		free(temp);
		list->size--;
	}

	return(data);
}


void *
list_remove_curr(LIST *ulist)
{
	list_element_t *temp;
	list_t *list = (list_t *)ulist;
	void *data;

	/* Remove the current element, returning a pointer to the data, or
	 * NULL if the list is empty.  Set curr to the next element unless
	 * curr is at the rear, in which case curr becomes the previous
	 * element.
	 */
	if (list->front == NULL) {
		/* The list is empty.  Easy. */
		return(NULL);
	} else if (list->front == list->rear) {
		/* The list has one element.  Easy. */
		data = list_remove_single(list);
	} else if (list->curr == list->front) {
		/* Removing front element.  Easy. */
		data = list_remove_front(list);
	} else if (list->curr == list->rear) {
		/* Removing from the rear.  Easy.*/
		data = list_remove_rear(list);
	} else {
		/* Otherwise.  Must be inside a 3-element or larger list. */
		data = list->curr->data;
		temp = list->curr;
		temp->next->prev = temp->prev;
		temp->prev->next = temp->next;
		list->curr = temp->next;
		free(temp);
		list->size--;
	}

	return(data);
}


int
list_traverse(LIST *ulist, void *data, list_comparator_t func, int opts)
{
	list_element_t *lp;
	list_t *list = (list_t *)ulist;
	int status, rc;

	/* Traverse the list according to opts, calling func at each element,
	 * until func returns 0 or the extent of the list is reached.  We return
	 * 0 if the list is empty, 2 if we tried to go beyond the extent of the
	 * list, and 1 otherwise.  We may or may not affect the current element
	 * pointer.
	 */
	if (list->front == NULL)
		return(LIST_EMPTY);
   
	/* Decide where to start. */
	if ((opts & LIST_CURR) == LIST_CURR) {
		lp = list->curr;
	} else if ((opts & LIST_REAR) == LIST_REAR) {
		lp = list->rear;
	} else {
		lp = list->front;
	}
   
	/* Now decide if to update the current element pointer. */
	if ((opts & LIST_ALTR) == LIST_ALTR)
		list->curr = lp;

	/* Now go until 0 is returned or we hit the extent of the list. */
	rc = LIST_OK;
	status = LIST_TRUE;
	while(status) {
		status = (*func)(data, lp->data);
		if (status) {
			if ((((opts & LIST_BACK) == LIST_BACK) ?
			     (lp->prev) : (lp->next)) == NULL) {
				/* Tried to go beyond extent of list. */
				status = LIST_FALSE;
				rc = LIST_EXTENT;
			} else {
				/* Decide where to go next. */
				lp = (((opts & LIST_BACK) == LIST_BACK) ?
				      (lp->prev) : (lp->next));

				/* Now decide if to update the current 
				 * element pointer. */
				if ((opts & LIST_ALTR) == LIST_ALTR)
					list->curr = lp;
			}
		}
	}

	return(rc);
}


